internal/sci/aws/server.go

```diff
@@ -0,0 +1,285 @@
+// Package aws provides an AWS implementation of the Substratus Cloud Interface (SCI)
+package aws
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	awsSdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/substratusai/substratus/internal/sci"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type Server struct {
+	sci.UnimplementedControllerServer
+	OIDCProviderURL string
+	OIDCProviderARN string
+	Clients
+}
+
+type Clients struct {
+	S3Client  *s3.S3
+	IamClient *iam.IAM
+}
+
+func NewAWSServer() (*Server, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create AWS session: %w", err)
+	}
+
+	clusterID, err := getClusterID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cluster ID: %w", err)
+	}
+
+	ec2Svc := ec2metadata.New(sess)
+	region, err := getRegion(ec2Svc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get region: %w", err)
+	}
+
+	stsSvc := sts.New(sess)
+	accountId, err := getAccountID(stsSvc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get account ID: %w", err)
+	}
+
+	OIDCProviderURL := fmt.Sprintf("oidc.eks.%s.amazonaws.com/id/%s", region, clusterID)
+	OIDCProviderARN := fmt.Sprintf("arn:aws:iam::%s:oidc-provider/%s", accountId, OIDCProviderURL)
+
+	c := &Clients{
+		S3Client:  s3.New(sess),
+		IamClient: iam.New(sess),
+	}
+
+	return &Server{
+		Clients:         *c,
+		OIDCProviderURL: OIDCProviderURL,
+		OIDCProviderARN: OIDCProviderARN,
+	}, nil
+}
+
+func (s *Server) GetObjectMd5(ctx context.Context, req *sci.GetObjectMd5Request) (*sci.GetObjectMd5Response, error) {
+	headResult, err := s.getObjectMetadata(s.Clients.S3Client, req.BucketName, req.ObjectName)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: AWS returns an MD5 checksum as an ETag except for multi-part uploads where it's an MD5 with a dash suffix.
+	if headResult.ETag == nil {
+		return nil, fmt.Errorf("object does not exist: %s", s3.ErrCodeNoSuchKey)
+	}
+
+	md5 := *headResult.ETag
+
+	return &sci.GetObjectMd5Response{
+		Md5Checksum: md5,
+	}, nil
+}
+
+func (s *Server) CreateSignedURL(ctx context.Context, req *sci.CreateSignedURLRequest) (*sci.CreateSignedURLResponse, error) {
+	bucketName, objectName, checksum := req.GetBucketName(),
+		req.GetObjectName(),
+		req.GetMd5Checksum()
+
+	// verify that the object doesn't exist
+	if _, err := s.getObjectMetadata(s.Clients.S3Client, bucketName, objectName); err != nil {
```

Comment:
Here's the other caller 😄

Reply:
`// verify that the object doesn't exist` -> Not following how this call is doing this. Btw, Is that desired behavior? How would a notebook re-upload to re-build?

Reply:
My confusion here. I think at some point we should do checks to be sure we aren't overwriting objects but not until we have good guarantees on path uniqueness (eg md5checksum as a part of the object key path)